OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs

```diff
@@ -34,68 +34,90 @@ public class WithAttackAnimationInfo : ITraitInfo, Requires<WithSpriteBodyInfo>,
+
+		public override object Create(ActorInitializer init) { return new WithAttackAnimation(init, this); }
+
+		public override void RulesetLoaded(Ruleset rules, ActorInfo ai)
+		{
+			var matches = ai.TraitInfos<WithSpriteBodyInfo>().Where(w => BodyName == w.Name);
+			if (matches.Count() != 1)
+				throw new YamlException("WithAttackAnimation needs exactly one sprite body with matching name.");
+
+			base.RulesetLoaded(rules, ai);
+		}
 	}
 
-	public class WithAttackAnimation : ITick, INotifyAttack
+	public class WithAttackAnimation : ConditionalTrait<WithAttackAnimationInfo>, ITick, INotifyAttack
 	{
-		readonly WithAttackAnimationInfo info;
 		readonly AttackBase attack;
 		readonly Armament armament;
 		readonly WithSpriteBody wsb;
+		readonly bool noAimOrReloadAnim;
 
 		int tick;
+		bool attackAnimPlaying;
 
 		public WithAttackAnimation(ActorInitializer init, WithAttackAnimationInfo info)
+			: base(info)
 		{
-			this.info = info;
 			attack = init.Self.Trait<AttackBase>();
 			armament = init.Self.TraitsImplementing<Armament>()
-				.Single(a => a.Info.Name == info.Armament);
-			wsb = init.Self.Trait<WithSpriteBody>();
+				.Single(a => a.Info.Name == Info.Armament);
+			wsb = init.Self.TraitsImplementing<WithSpriteBody>().First(w => Info.BodyName == w.Info.Name);
+
+			noAimOrReloadAnim = string.IsNullOrEmpty(Info.AimSequence) && string.IsNullOrEmpty(Info.ReloadPrefix);
 		}
 
 		void PlayAttackAnimation(Actor self)
 		{
-			if (!string.IsNullOrEmpty(info.AttackSequence))
-				wsb.PlayCustomAnimation(self, info.AttackSequence, () => wsb.CancelCustomAnimation(self));
+			if (!IsTraitDisabled && wsb.IsTraitDisabled && !string.IsNullOrEmpty(Info.AttackSequence))
+			{
+				attackAnimPlaying = true;
+				wsb.PlayCustomAnimation(self, Info.AttackSequence,
+					() => { wsb.CancelCustomAnimation(self); attackAnimPlaying = false; });
+			}
 		}
 
 		void INotifyAttack.Attacking(Actor self, Target target, Armament a, Barrel barrel)
 		{
-			if (info.DelayRelativeTo == AttackDelayType.Attack)
+			if (Info.DelayRelativeTo == AttackDelayType.Attack)
 			{
-				if (info.Delay > 0)
-					tick = info.Delay;
+				if (Info.Delay > 0)
+					tick = Info.Delay;
 				else
 					PlayAttackAnimation(self);
 			}
 		}
 
 		void INotifyAttack.PreparingAttack(Actor self, Target target, Armament a, Barrel barrel)
 		{
-			if (info.DelayRelativeTo == AttackDelayType.Preparation)
+			if (Info.DelayRelativeTo == AttackDelayType.Preparation)
 			{
-				if (info.Delay > 0)
-					tick = info.Delay;
+				if (Info.Delay > 0)
+					tick = Info.Delay;
 				else
 					PlayAttackAnimation(self);
 			}
 		}
 
 		void ITick.Tick(Actor self)
 		{
-			if (info.Delay > 0 && --tick == 0)
+			if (Info.Delay > 0 && --tick == 0)
 				PlayAttackAnimation(self);
 
-			if (string.IsNullOrEmpty(info.AimSequence) && string.IsNullOrEmpty(info.ReloadPrefix))
+			if (IsTraitDisabled || noAimOrReloadAnim || attackAnimPlaying)
+				return;
+
+			if (wsb.IsTraitDisabled)
```

Comment:
I'm on the fence about whether this should be included above. `IsTraitDisabled || noAimOrReloadAnim || attackAnimPlaying || wsb.IsTraitDisabled`